File: internal/proto/fs.go
package proto

// FSEntryType 标识目录项类型。
type FSEntryType string

const (
	FSEntryFile    FSEntryType = "file"
	FSEntryDir     FSEntryType = "dir"
	FSEntrySymlink FSEntryType = "symlink"
)

// FSEntry 是目录列表里的一项。
type FSEntry struct {
	Name    string      `json:"name"`
	Type    FSEntryType `json:"type"`
	Size    int64       `json:"size"`
	ModTime int64       `json:"mod_time"` // unix ms
}

// 文件服务 op 常量。
const (
	OpFSList   = "fs.list"
	OpFSRead   = "fs.read"
	OpFSWrite  = "fs.write"
	OpFSStat   = "fs.stat"
	OpFSMkdir  = "fs.mkdir"
	OpFSRemove = "fs.remove"
	OpFSMove   = "fs.move"
)

// FSListReq / FSListRes。
type FSListReq struct {
	Path string `json:"path"` // 客户端本地路径；为 "." 时由客户端解析为工作目录
}
type FSListRes struct {
	Path    string    `json:"path"` // 服务端解析后的绝对路径
	Entries []FSEntry `json:"entries"`
}

// FSReadReq / FSReadRes。
// Encoding 为 "utf8" 时 Content 是 UTF-8 文本；为 "base64" 时是 base64 编码的二进制。
type FSReadReq struct {
	Path string `json:"path"`
}
type FSReadRes struct {
	Encoding string `json:"encoding"`
	Content  string `json:"content"`
	Size     int64  `json:"size"`
}

// FSWriteReq 把内容写入指定路径。
type FSWriteReq struct {
	Path     string `json:"path"`
	Encoding string `json:"encoding"` // "utf8" 或 "base64"
	Content  string `json:"content"`
}
type FSWriteRes struct {
	Size int64 `json:"size"`
}

// FSStatReq / FSStatRes。
type FSStatReq struct {
	Path string `json:"path"`
}
type FSStatRes struct {
	Type    FSEntryType `json:"type"`
	Size    int64       `json:"size"`
	ModTime int64       `json:"mod_time"`
}

// FSMkdirReq 创建目录（含父目录）。
type FSMkdirReq struct {
	Path string `json:"path"`
}

// FSRemoveReq 删除文件或目录。
type FSRemoveReq struct {
	Path      string `json:"path"`
	Recursive bool   `json:"recursive"`
}

// FSMoveReq 移动 / 重命名。
type FSMoveReq struct {
	From string `json:"from"`
	To   string `json:"to"`
}

// FSError 是所有 fs op 失败时的统一响应体。
type FSError struct {
	Error string `json:"error"`
}

File: web/src/api/fs.ts
// fs.ts —— 浏览器侧 fs API 客户端封装。所有路径都走 query string。
// 路径区分：
//   - 显示路径：来自 list response.path（绝对路径，POSIX-style）
//   - 请求路径：发给 client 的本地路径，Windows 上是反斜杠（见下）

// list response.path 是 POSIX-style（filepath.ToSlash 结果），但发回 client 时
// 需要原样（client 端 filepath.Abs 兼容）。Windows 下 ToSlash 把 "C:\" 变成 "C:/"，
// filepath.Abs("C:/foo") 仍能正确解析，因此往返一致。

export type FSEntryType = "file" | "dir" | "symlink";

export type FSEntry = {
  name: string;
  type: FSEntryType;
  size: number;
  mod_time: number;
};

export type FSListRes = {
  path: string;
  entries: FSEntry[];
};

export type FSReadRes = {
  encoding: "utf8" | "base64";
  content: string;
  size: number;
};

async function jsonOrThrow(r: Response): Promise<any> {
  if (!r.ok) {
    let msg = `${r.status}`;
    try {
      const j = await r.json();
      if (j && j.error) msg = j.error;
    } catch {}
    throw new Error(msg);
  }
  return r.json();
}

export async function fsList(clientId: string, path: string): Promise<FSListRes> {
  const r = await fetch(
    `/api/fs/${encodeURIComponent(clientId)}/list?path=${encodeURIComponent(path)}`,
    { credentials: "include" },
  );
  return jsonOrThrow(r);
}

export async function fsRead(clientId: string, path: string): Promise<FSReadRes> {
  const r = await fetch(
    `/api/fs/${encodeURIComponent(clientId)}/read?path=${encodeURIComponent(path)}`,
    { credentials: "include" },
  );
  return jsonOrThrow(r);
}

export async function fsWrite(
  clientId: string,
  path: string,
  content: string,
  encoding: "utf8" | "base64" = "utf8",
): Promise<{ size: number }> {
  const r = await fetch(
    `/api/fs/${encodeURIComponent(clientId)}/write`,
    {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ path, content, encoding }),
    },
  );
  return jsonOrThrow(r);
}

export async function fsMkdir(clientId: string, path: string): Promise<void> {
  const r = await fetch(
    `/api/fs/${encodeURIComponent(clientId)}/mkdir?path=${encodeURIComponent(path)}`,
    { method: "POST", credentials: "include" },
  );
  await jsonOrThrow(r);
}

export async function fsRemove(
  clientId: string,
  path: string,
  recursive = false,
): Promise<void> {
  const r = await fetch(
    `/api/fs/${encodeURIComponent(clientId)}/remove?path=${encodeURIComponent(path)}&recursive=${recursive}`,
    { method: "POST", credentials: "include" },
  );
  await jsonOrThrow(r);
}
